File: go/fory/decimal.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package fory

import (
	"fmt"
	"math/big"
	"reflect"
)

// Decimal is the Go carrier for Fory DECIMAL values.
// It preserves the exact value unscaled * 10^-scale.
type Decimal struct {
	Unscaled big.Int
	Scale    int32
}

func NewDecimal(unscaled *big.Int, scale int32) Decimal {
	var copied big.Int
	if unscaled != nil {
		copied.Set(unscaled)
	}
	return Decimal{Unscaled: copied, Scale: scale}
}

func (d Decimal) Equal(other Decimal) bool {
	return d.Scale == other.Scale && d.Unscaled.Cmp(&other.Unscaled) == 0
}

func (d Decimal) String() string {
	return fmt.Sprintf("%se%d", d.Unscaled.String(), -d.Scale)
}

var (
	decimalReflectType = reflect.TypeFor[Decimal]()
	decimalLongMin     = big.NewInt(MinInt64)
	decimalLongMax     = big.NewInt(MaxInt64)
)

type decimalSerializer struct{}

func (s decimalSerializer) Write(ctx *WriteContext, refMode RefMode, writeType bool, hasGenerics bool, value reflect.Value) {
	if refMode != RefModeNone {
		ctx.buffer.WriteInt8(NotNullValueFlag)
	}
	if writeType {
		ctx.buffer.WriteUint8(uint8(DECIMAL))
	}
	s.WriteData(ctx, value)
}

func (s decimalSerializer) WriteData(ctx *WriteContext, value reflect.Value) {
	decimal := value.Interface().(Decimal)
	writeDecimalParts(ctx.buffer, decimal.Scale, &decimal.Unscaled)
}

func (s decimalSerializer) Read(ctx *ReadContext, refMode RefMode, readType bool, hasGenerics bool, value reflect.Value) {
	err := ctx.Err()
	if refMode != RefModeNone {
		if ctx.buffer.ReadInt8(err) == NullFlag {
			value.Set(reflect.Zero(value.Type()))
			return
		}
	}
	if readType && !ctx.readExpectedTypeID(DECIMAL) {
		return
	}
	if ctx.HasError() {
		return
	}
	s.ReadData(ctx, value)
}

func (s decimalSerializer) ReadData(ctx *ReadContext, value reflect.Value) {
	scale, unscaled := readDecimalParts(ctx)
	if ctx.HasError() {
		return
	}
	value.Set(reflect.ValueOf(NewDecimal(unscaled, scale)))
}

func (s decimalSerializer) ReadWithTypeInfo(ctx *ReadContext, refMode RefMode, typeInfo *TypeInfo, value reflect.Value) {
	s.Read(ctx, refMode, false, false, value)
}

func writeDecimalParts(buffer *ByteBuffer, scale int32, unscaled *big.Int) {
	if unscaled == nil {
		unscaled = new(big.Int)
	}
	buffer.WriteVarint32(scale)
	if canUseSmallDecimalEncoding(unscaled) {
		smallValue := unscaled.Int64()
		header := encodeDecimalZigZag64(smallValue) << 1
		buffer.WriteVarUint64(header)
		return
	}

	abs := new(big.Int).Abs(unscaled)
	payload := abs.Bytes()
	if len(payload) == 0 {
		panic("decimal zero must use the small encoding")
	}
	reverseBytes(payload)
	meta := (uint64(len(payload)) << 1) | uint64(signBit(unscaled.Sign()))
	buffer.WriteVarUint64((meta << 1) | 1)
	buffer.WriteBinary(payload)
}

func readDecimalParts(ctx *ReadContext) (int32, *big.Int) {
	err := ctx.Err()
	scale := ctx.buffer.ReadVarint32(err)
	header := ctx.buffer.ReadVarUint64(err)
	if ctx.HasError() {
		return 0, nil
	}
	if (header & 1) == 0 {
		return scale, big.NewInt(decodeDecimalZigZag64(header >> 1))
	}

	meta := header >> 1
	length := meta >> 1
	if length == 0 {
		ctx.SetError(DeserializationErrorf("invalid decimal magnitude length %d", length))
		return 0, nil
	}
	if length > uint64(ctx.maxBinarySize) {
		ctx.SetError(MaxBinarySizeExceededError(int(length), ctx.maxBinarySize))
		return 0, nil
	}
	payload := ctx.buffer.ReadBytes(int(length), err)
	if ctx.HasError() {
		return 0, nil
	}
	if payload[len(payload)-1] == 0 {
		ctx.SetError(DeserializationError("non-canonical decimal payload: trailing zero byte"))
		return 0, nil
	}
	bigEndian := append([]byte(nil), payload...)
	reverseBytes(bigEndian)
	magnitude := new(big.Int).SetBytes(bigEndian)
	if magnitude.Sign() == 0 {
		ctx.SetError(DeserializationError("big decimal encoding must not represent zero"))
		return 0, nil
	}
	if (meta & 1) != 0 {
		magnitude.Neg(magnitude)
	}
	return scale, magnitude
}

func canUseSmallDecimalEncoding(value *big.Int) bool {
	if value == nil {
		return true
	}
	if value.Cmp(decimalLongMin) < 0 || value.Cmp(decimalLongMax) > 0 {
		return false
	}
	return (encodeDecimalZigZag64(value.Int64()) & (1 << 63)) == 0
}

func encodeDecimalZigZag64(value int64) uint64 {
	return uint64((value << 1) ^ (value >> 63))
}

func decodeDecimalZigZag64(value uint64) int64 {
	return int64((value >> 1) ^ uint64(-(int64(value & 1))))
}

func signBit(sign int) int {
	if sign < 0 {
		return 1
	}
	return 0
}

func reverseBytes(values []byte) {
	for i, j := 0, len(values)-1; i < j; i, j = i+1, j-1 {
		values[i], values[j] = values[j], values[i]
	}
}

File: rust/fory-core/src/serializer/decimal.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use crate::buffer::{Reader, Writer};
use crate::context::{ReadContext, WriteContext};
use crate::error::Error;
use crate::resolver::TypeResolver;
use crate::serializer::util::read_basic_type_info;
use crate::serializer::{ForyDefault, Serializer};
use crate::type_id::TypeId;
use crate::types::Decimal;
use num_bigint::{BigInt, Sign};
use std::convert::TryFrom;

impl Serializer for Decimal {
    #[inline(always)]
    fn fory_write_data(&self, context: &mut WriteContext) -> Result<(), Error> {
        context.writer.write_var_i32(self.scale);
        write_decimal_unscaled(&self.unscaled, &mut context.writer)
    }

    #[inline(always)]
    fn fory_read_data(context: &mut ReadContext) -> Result<Self, Error> {
        let scale = context.reader.read_var_i32()?;
        let unscaled = read_decimal_unscaled(&mut context.reader)?;
        Ok(Self { unscaled, scale })
    }
    #[inline]
    fn fory_read_data_as_send_sync_any(
        context: &mut ReadContext,
    ) -> Result<Box<dyn std::any::Any + Send + Sync>, Error>
    where
        Self: Sized + ForyDefault,
    {
        Ok(crate::serializer::box_send_sync(Self::fory_read_data(
            context,
        )?))
    }

    #[inline(always)]
    fn fory_get_type_id(_: &TypeResolver) -> Result<TypeId, Error> {
        Ok(TypeId::DECIMAL)
    }

    #[inline(always)]
    fn fory_type_id_dyn(&self, _: &TypeResolver) -> Result<TypeId, Error> {
        Ok(TypeId::DECIMAL)
    }

    #[inline(always)]
    fn fory_static_type_id() -> TypeId {
        TypeId::DECIMAL
    }

    #[inline(always)]
    fn as_any(&self) -> &dyn std::any::Any {
        self
    }

    #[inline(always)]
    fn fory_write_type_info(context: &mut WriteContext) -> Result<(), Error> {
        context.writer.write_var_u32(TypeId::DECIMAL as u32);
        Ok(())
    }

    #[inline(always)]
    fn fory_read_type_info(context: &mut ReadContext) -> Result<(), Error> {
        read_basic_type_info::<Self>(context)
    }
}

impl ForyDefault for Decimal {
    #[inline(always)]
    fn fory_default() -> Self {
        Self {
            unscaled: BigInt::from(0),
            scale: 0,
        }
    }
}

fn write_decimal_unscaled(value: &BigInt, writer: &mut Writer) -> Result<(), Error> {
    if let Some(small_value) = can_use_small_encoding(value) {
        writer.write_var_u64(encode_zigzag64(small_value) << 1);
        return Ok(());
    }

    let (sign, payload) = value.to_bytes_le();
    if payload.is_empty() {
        return Err(Error::invalid_data(
            "zero must use the small decimal encoding".to_string(),
        ));
    }
    let meta = ((payload.len() as u64) << 1) | u64::from(matches!(sign, Sign::Minus));
    writer.write_var_u64((meta << 1) | 1);
    writer.write_bytes(&payload);
    Ok(())
}

fn read_decimal_unscaled(reader: &mut Reader) -> Result<BigInt, Error> {
    let header = reader.read_var_u64()?;
    if (header & 1) == 0 {
        return Ok(BigInt::from(decode_zigzag64(header >> 1)));
    }

    let meta = header >> 1;
    let sign = (meta & 1) != 0;
    let len = (meta >> 1) as usize;
    if len == 0 {
        return Err(Error::invalid_data(
            "invalid decimal magnitude length 0".to_string(),
        ));
    }
    let payload = reader.read_bytes(len)?;
    if payload[len - 1] == 0 {
        return Err(Error::invalid_data(
            "non-canonical decimal payload: trailing zero byte".to_string(),
        ));
    }
    let magnitude = BigInt::from_bytes_le(Sign::Plus, payload);
    if magnitude == BigInt::from(0) {
        return Err(Error::invalid_data(
            "big decimal encoding must not represent zero".to_string(),
        ));
    }
    Ok(if sign { -magnitude } else { magnitude })
}

fn can_use_small_encoding(value: &BigInt) -> Option<i64> {
    let small_value = i64::try_from(value).ok()?;
    if (encode_zigzag64(small_value) & (1u64 << 63)) == 0 {
        Some(small_value)
    } else {
        None
    }
}

#[inline(always)]
fn encode_zigzag64(value: i64) -> u64 {
    ((value << 1) ^ (value >> 63)) as u64
}

#[inline(always)]
fn decode_zigzag64(value: u64) -> i64 {
    ((value >> 1) as i64) ^ -((value & 1) as i64)
}
